File: VB/DevExpress.XtraReports.CustomControls/XRTreeListPresenters.vb
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Reflection
Imports System.Text
Imports DevExpress.Utils.Serializing
Imports DevExpress.XtraPrinting
Imports DevExpress.XtraPrinting.BrickExporters
Imports DevExpress.XtraPrinting.Export
Imports DevExpress.XtraPrinting.Native
Imports DevExpress.XtraReports.Native.Presenters
Imports DevExpress.XtraReports.Native.Printing
Imports DevExpress.XtraReports.UI

Namespace DevExpress.XtraReports.CustomControls
	Friend Class XRTreeListRuntimePresenter
		Inherits XRTableLikeContainerControlPresenter

		#Region "Methods"
		Public Sub New(ByVal treeList As XRTreeList)
			MyBase.New(treeList)
		End Sub

		Protected Overrides Function GetActualDataCollection() As XRDataRecordCollection
			Return TreeList.Nodes
		End Function

		Protected Overrides Sub CreateBricksForRecord(ByVal record As XRDataRecord, ByVal parentBrick As PanelBrick, ByVal isHeader As Boolean, ByRef actualHeight As Single)
			Dim node As XRTreeListNode = TryCast(record, XRTreeListNode)
			Dim args As New PrintNodeEventArgs(node)
			If Not isHeader Then
				TreeList.OnPrintNode(args)
			End If

			If args.SuppressType = NodeSuppressType.Leave Then
				Dim nodeLevel As Integer = node.Level

				Dim nodeBrick As New TreeListNodeBrick(TreeList, DirectCast(parentBrick, DataContainerBrick), isHeader)
				nodeBrick.Style = TryCast(XRControlStyle.Default.Clone(), XRControlStyle)
				nodeBrick.Separable = False

				Dim cache As RecordPrintCache = New TreeListNodePrintCache(nodeBrick, node.Level)

				Dim childBricks As New List(Of VisualBrick)()
				Dim visibleHeaders As List(Of XRFieldHeader) = TreeList.VisibleHeaders

				For i As Integer = 0 To visibleHeaders.Count - 1
					Dim valueBrick As VisualBrick = CreateCellBrick(TreeList, parentBrick, node, i, isHeader)
					childBricks.Add(valueBrick)
				Next i

				CorrectBrickBounds(nodeBrick, childBricks, nodeLevel * TreeList.NodeIndent, actualHeight)
				Dim nodeHeight As Single = nodeBrick.Rect.Height

				VisualBrickHelper.SetBrickBounds(nodeBrick, nodeBrick.Rect, TreeList.Dpi)

				parentBrick.Bricks.Add(nodeBrick)

				actualHeight += nodeHeight

				If Not IsDesignMode Then
					If isHeader Then
						DirectCast(parentBrick, DataContainerBrick).PrintCache.HeaderCache = cache
					Else
						DirectCast(parentBrick, DataContainerBrick).PrintCache.RecordsCache.Add(cache)
					End If
				End If
			End If

			If args.SuppressType <> NodeSuppressType.SuppressWithChildren Then
				For Each childNode As XRTreeListNode In node.Nodes
					CreateBricksForRecord(childNode, parentBrick, isHeader, actualHeight)
				Next childNode
			End If
		End Sub

		Protected Overrides Function CreateBrickStyle(ByVal control As XRDataContainerControl, ByVal parentBrick As VisualBrick, ByVal valueBrick As VisualBrick, ByVal record As XRDataRecord, ByVal fieldIndex As Integer, ByVal isHeader As Boolean) As BrickStyle
			Dim style As BrickStyle = MyBase.CreateBrickStyle(control, parentBrick, valueBrick, record, fieldIndex, isHeader)

			If Not isHeader Then
				Dim printCellArgs As PrintCellEventArgs = New PrintNodeCellEventArgs(DirectCast(record, XRTreeListNode), DirectCast(control.VisibleHeaders(fieldIndex), XRTreeListColumn), valueBrick, style)
				TreeList.OnPrintNodeCell(DirectCast(printCellArgs, PrintNodeCellEventArgs))
			End If

			Return style
		End Function

		Protected Overrides Function CreateCellBrick(ByVal control As XRDataContainerControl, ByVal parentBrick As VisualBrick, ByVal record As XRDataRecord, ByVal fieldIndex As Integer, ByVal isHeader As Boolean) As VisualBrick
			Dim valueBrick As VisualBrick = MyBase.CreateCellBrick(control, parentBrick, record, fieldIndex, isHeader)

			Dim columnWidth As Single = DirectCast(control.VisibleHeaders(fieldIndex), XRTreeListColumn).Width
			Dim columnPos As Single = 0

			If fieldIndex > 0 Then
				columnPos -= DirectCast(record, XRTreeListNode).Level * TreeList.NodeIndent
			End If

			For j As Integer = 0 To fieldIndex - 1
				columnPos += DirectCast(control.VisibleHeaders(j), XRTreeListColumn).Width
			Next j

			Dim curColWidth As Single = If(fieldIndex = 0, columnWidth - DirectCast(record, XRTreeListNode).Level * TreeList.NodeIndent, columnWidth)

			valueBrick.Style = CreateBrickStyle(control, parentBrick, valueBrick, record, fieldIndex, isHeader)

			Dim brickHeight As Single = GetBrickHeight(valueBrick, columnWidth, isHeader)

			valueBrick.Rect = New RectangleF(columnPos, 0, curColWidth, brickHeight)

			Return valueBrick
		End Function
		#End Region

		#Region "Properties"
		Protected ReadOnly Property TreeList() As XRTreeList
			Get
                Return DirectCast(control, XRTreeList)
			End Get
		End Property
		#End Region
	End Class

	Friend Class XRTreeListDesignTimePresenter
		Inherits XRTreeListRuntimePresenter

		#Region "Methods"
		Public Sub New(ByVal treeList As XRTreeList)
			MyBase.New(treeList)
		End Sub

		Public Overrides Function CreateBrick(ByVal childrenBricks() As VisualBrick) As VisualBrick
			Return New DataContainerBrick(TreeList, False) With {.PrintCache = New XRDataContainerPrintCache(DirectCast(Control, XRTreeList))}
		End Function

		Protected Overrides Function GetActualDataCollection() As XRDataRecordCollection
			Return CreateDesignNodes()
		End Function

		Private Function CreateDesignNodes() As XRTreeListNodeCollection
			Dim designNodes As New XRTreeListNodeCollection(Nothing)

			Dim parentNode As XRTreeListNode = Nothing

			For i As Integer = 0 To 2
				Dim currentNode As New XRTreeListNode(TreeList)

				Dim visibleHeaders As List(Of XRFieldHeader) = TreeList.VisibleHeaders

				For j As Integer = 0 To visibleHeaders.Count - 1
					If visibleHeaders(j).FieldType IsNot Nothing Then
						currentNode(j) = visibleHeaders(j).FieldType.Name
					End If
				Next j

				If parentNode Is Nothing Then
					designNodes.Add(currentNode)
					parentNode = currentNode
				Else
					parentNode.AddNode(currentNode)
					parentNode = currentNode
				End If
			Next i

			Return designNodes
		End Function

		#End Region

		#Region "Properties"
		Protected Overrides ReadOnly Property IsDesignMode() As Boolean
			Get
				Return True
			End Get
		End Property
		#End Region
	End Class
End Namespace

File: CS/DevExpress.XtraReports.CustomControls/XRTreeListPresenters.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Reflection;
using System.Text;
using DevExpress.Utils.Serializing;
using DevExpress.XtraPrinting;
using DevExpress.XtraPrinting.BrickExporters;
using DevExpress.XtraPrinting.Export;
using DevExpress.XtraPrinting.Native;
using DevExpress.XtraReports.Native.Presenters;
using DevExpress.XtraReports.Native.Printing;
using DevExpress.XtraReports.UI;

namespace DevExpress.XtraReports.CustomControls
{
    internal class XRTreeListRuntimePresenter : XRTableLikeContainerControlPresenter
    {
        public XRTreeListRuntimePresenter(XRTreeList treeList) : base(treeList)
        {
        }

        protected override XRDataRecordCollection GetActualDataCollection()
        {
            return TreeList.Nodes;
        }

        protected override void CreateBricksForRecord(XRDataRecord record, PanelBrick parentBrick, bool isHeader, ref float actualHeight)
        {
            XRTreeListNode node = record as XRTreeListNode;
            PrintNodeEventArgs args = new PrintNodeEventArgs(node);
            if (!isHeader)
                TreeList.OnPrintNode(args);

            if (args.SuppressType == NodeSuppressType.Leave)
            {
                int nodeLevel = node.Level;

                TreeListNodeBrick nodeBrick = new TreeListNodeBrick(TreeList, (DataContainerBrick)parentBrick, isHeader);
                nodeBrick.Style = XRControlStyle.Default.Clone() as XRControlStyle;
                nodeBrick.Separable = false;

                RecordPrintCache cache = new TreeListNodePrintCache(nodeBrick, node.Level);

                List<VisualBrick> childBricks = new List<VisualBrick>();
                List<XRFieldHeader> visibleHeaders = TreeList.VisibleHeaders;

                for (int i = 0; i < visibleHeaders.Count; i++)
                {
                    VisualBrick valueBrick = CreateCellBrick(TreeList, parentBrick, node, i, isHeader);
                    childBricks.Add(valueBrick);
                }

                CorrectBrickBounds(nodeBrick, childBricks, nodeLevel * TreeList.NodeIndent, actualHeight);
                float nodeHeight = nodeBrick.Rect.Height;

                VisualBrickHelper.SetBrickBounds(nodeBrick, nodeBrick.Rect, TreeList.Dpi);

                parentBrick.Bricks.Add(nodeBrick);

                actualHeight += nodeHeight;

                if (!IsDesignMode)
                    if (isHeader)
                        ((DataContainerBrick)parentBrick).PrintCache.HeaderCache = cache;
                    else
                        ((DataContainerBrick)parentBrick).PrintCache.RecordsCache.Add(cache);
            }

            if (args.SuppressType != NodeSuppressType.SuppressWithChildren)
                foreach (XRTreeListNode childNode in node.Nodes)
                    CreateBricksForRecord(childNode, parentBrick, isHeader, ref actualHeight);
        }

        protected override BrickStyle CreateBrickStyle(XRDataContainerControl control, VisualBrick parentBrick, VisualBrick valueBrick, XRDataRecord record, int fieldIndex, bool isHeader)
        {
            BrickStyle style = base.CreateBrickStyle(control, parentBrick, valueBrick, record, fieldIndex, isHeader);

            if (!isHeader)
            {
                PrintCellEventArgs printCellArgs = new PrintNodeCellEventArgs((XRTreeListNode)record, (XRTreeListColumn)control.VisibleHeaders[fieldIndex], valueBrick, style);
                TreeList.OnPrintNodeCell((PrintNodeCellEventArgs)printCellArgs);
            }

            return style;
        }

        protected override VisualBrick CreateCellBrick(XRDataContainerControl control, VisualBrick parentBrick, XRDataRecord record, int fieldIndex, bool isHeader)
        {
            VisualBrick valueBrick = base.CreateCellBrick(control, parentBrick, record, fieldIndex, isHeader);

            float columnWidth = ((XRTreeListColumn)control.VisibleHeaders[fieldIndex]).Width;
            float columnPos = 0;

            if (fieldIndex > 0)
                columnPos -= ((XRTreeListNode)record).Level * TreeList.NodeIndent;

            for (int j = 0; j < fieldIndex; j++)
                columnPos += ((XRTreeListColumn)control.VisibleHeaders[j]).Width;

            float curColWidth = fieldIndex == 0 ? columnWidth - ((XRTreeListNode)record).Level * TreeList.NodeIndent : columnWidth;                                    

            valueBrick.Style = CreateBrickStyle(control, parentBrick, valueBrick, record, fieldIndex, isHeader);

            float brickHeight = GetBrickHeight(valueBrick, columnWidth, isHeader);

            valueBrick.Rect = new RectangleF(columnPos, 0, curColWidth, brickHeight);

            return valueBrick;
        }

        protected XRTreeList TreeList
        {
            get { return (XRTreeList)control; }
        }
    }

    internal class XRTreeListDesignTimePresenter : XRTreeListRuntimePresenter
    {
        public XRTreeListDesignTimePresenter(XRTreeList treeList) : base(treeList) { }

		public override VisualBrick CreateBrick(VisualBrick[] childrenBricks) {
            return new DataContainerBrick(TreeList, false) { PrintCache = new XRDataContainerPrintCache((XRTreeList)ContainerControl) };
		}

        protected override XRDataRecordCollection GetActualDataCollection()
        {
            return CreateDesignNodes();
        }

        private XRTreeListNodeCollection CreateDesignNodes()
        {
            XRTreeListNodeCollection designNodes = new XRTreeListNodeCollection(null);

            XRTreeListNode parentNode = null;

            for (int i = 0; i < 3; i++)
            {
                XRTreeListNode currentNode = new XRTreeListNode(TreeList);

                List<XRFieldHeader> visibleHeaders = TreeList.VisibleHeaders;

                for (int j = 0; j < visibleHeaders.Count; j++)
                    if (visibleHeaders[j].FieldType != null)
                        currentNode[j] = visibleHeaders[j].FieldType.Name;

                if (parentNode == null)
                {
                    designNodes.Add(currentNode);
                    parentNode = currentNode;
                }
                else
                {
                    parentNode.AddNode(currentNode);
                    parentNode = currentNode;
                }
            }

            return designNodes;
        }


        protected override bool IsDesignMode { get { return true; } }
    }
}
